web3/middleware/cache.py

```diff
@@ -0,0 +1,359 @@
+    # 'eth_getBlockTransactionCountByNumber',
+    # 'eth_getUncleCountByBlockHash',
+    # 'eth_getUncleCountByBlockNumber',
+    # 'eth_getCode',
+    # 'eth_sign',
+    # 'eth_sendTransaction',
+    # 'eth_sendRawTransaction',
+    # 'eth_call',
+    # 'eth_estimateGas',
+    # 'eth_getBlockByHash',
+    # 'eth_getBlockByNumber',
+    # 'eth_getTransactionByHash',
+    # 'eth_getTransactionByBlockHashAndIndex',
+    # 'eth_getTransactionByBlockNumberAndIndex',
+    # 'eth_getTransactionReceipt',
+    # 'eth_getUncleByBlockHashAndIndex',
+    # 'eth_getUncleByBlockNumberAndIndex',
+    # 'eth_getCompilers',
+    # 'eth_compileLLL',
+    # 'eth_compileSolidity',
+    # 'eth_compileSerpent',
+    # 'eth_newFilter',
+    # 'eth_newBlockFilter',
+    # 'eth_newPendingTransactionFilter',
+    # 'eth_uninstallFilter',
+    # 'eth_getFilterChanges',
+    # 'eth_getFilterLogs',
+    # 'eth_getLogs',
+    # 'eth_getWork',
+    # 'eth_submitWork',
+    # 'eth_submitHashrate',
+}
+
+
+def construct_time_based_cache_middleware(cache,
+                                          cache_expire_seconds=15,
+                                          rpc_whitelist=TIME_BASED_CACHE_RPC_WHITELIST,
+                                          should_cache_fn=should_cache):
+    """
+    Constructs a middleware which caches responses based on the request
+    ``method`` and ``params`` for a maximum amount of time as specified
+
+    :param cache: Any dictionary-like object
+    :param cache_expire_seconds: The number of seconds an item may be cached before it should expire.
+    :param rpc_whitelist: A set of RPC methods which may have their responses cached.
+    :param should_cache_fn: A callable which accepts ``method`` ``params`` and ``response`` and returns a boolean as to whether the response should be cached.
+    """
+    def simple_cache_middleware(make_request, web3):
+        def middleware(method, params):
+            if method in rpc_whitelist:
+                cache_key = generate_cache_key((method, params))
+                if cache_key in cache:
+                    # check that the cached response is not expired.
+                    cached_at, cached_response = cache[cache_key]
+                    cached_for = time.time() - cached_at
+
+                    if cached_for <= cache_expire_seconds:
+                        return cached_response
+                    else:
+                        del cache[cache_key]
+
+                # cache either missed or expired so make the request.
+                response = make_request(method, params)
+
+                if should_cache_fn(response):
+                    cache[cache_key] = (time.time(), response)
+
+                return response
+            else:
+                return make_request(method, params)
+        return middleware
+    return simple_cache_middleware
+
+
+time_based_cache_middleware = construct_time_based_cache_middleware(
+    cache=lru.LRU(256),
+)
```

Comment:
I'm a compulsive double-checker. So I'll almost always go and request `eth_accounts` immediately after creating an account. Not seeing the new account would feel broken here, and 15s is kind of a long time in that context. Short enough to be reasonable in that use case also means too short to be of much good, I would think.

As long as this cache doesn't become a default, I suppose it's not a problem, though.

Reply:
I don't think any of these should ever be default so I think we're ok there.